src/wasm-type.h

```diff
@@ -78,6 +78,227 @@ using HeapTypeNameGenerator = std::function<TypeNames(HeapType)>;
+  // number of supertypes in its supertype chain.
+  size_t getDepth() const;
+
+  // Get the bottom heap type for this heap type's hierarchy.
+  BasicHeapType getUnsharedBottom() const;
+  BasicHeapType getBottom() const {
+    return HeapType(getUnsharedBottom()).getBasic(getShared());
+  }
+
+  // Get the top heap type for this heap type's hierarchy.
+  BasicHeapType getUnsharedTop() const;
+  BasicHeapType getTop() const {
+    return HeapType(getUnsharedTop()).getBasic(getShared());
+  }
+
+  // Get the recursion group for this non-basic type.
+  RecGroup getRecGroup() const;
+
+  // Get the index of this non-basic type within its recursion group.
+  size_t getRecGroupIndex() const;
+
+  constexpr TypeID getID() const { return id; }
+
+  // Get the shared or unshared version of this basic heap type.
+  constexpr BasicHeapType getBasic(Shareability share) const {
+    assert(isBasic());
+    return BasicHeapType(share == Shared ? (id | 1) : (id & ~1));
+  }
+
+  // (In)equality must be defined for both HeapType and BasicHeapType because it
+  // is otherwise ambiguous whether to convert both this and other to int or
+  // convert other to HeapType.
+  bool operator==(const HeapType& other) const { return id == other.id; }
+  bool operator==(const BasicHeapType& other) const { return id == other; }
+  bool operator!=(const HeapType& other) const { return id != other.id; }
+  bool operator!=(const BasicHeapType& other) const { return id != other; }
+
+  // Returns true if left is a subtype of right. Subtype includes itself.
+  static bool isSubType(HeapType left, HeapType right);
+
+  std::vector<Type> getTypeChildren() const;
+
+  // Return the ordered HeapType children, looking through child Types.
+  std::vector<HeapType> getHeapTypeChildren() const;
+
+  // Similar to `getHeapTypeChildren`, but also includes the supertype if it
+  // exists.
+  std::vector<HeapType> getReferencedHeapTypes() const;
+
+  // Return the LUB of two HeapTypes, which may or may not exist.
+  static std::optional<HeapType> getLeastUpperBound(HeapType a, HeapType b);
+
+  // Returns the feature set required to use this type.
+  FeatureSet getFeatures() const;
+
+  // Helper allowing the value of `print(...)` to be sent to an ostream. Stores
+  // a `TypeID` because `Type` is incomplete at this point and using a reference
+  // makes it less convenient to use.
+  struct Printed {
+    TypeID typeID;
+    HeapTypeNameGenerator generateName;
+  };
+
+  // Given a function for generating HeapType names, print the definition of
+  // this HeapType to `os`. `generateName` should return the same
+  // name each time it is called with the same HeapType and it should return
+  // different names for different types.
+  Printed print(HeapTypeNameGenerator generateName) {
+    return Printed{getID(), generateName};
+  }
+
+  std::string toString() const;
+};
+
+// Internal only.
+struct TypeInfo {
```

Comment:
Can we put this and the method implementations that use it into a separate `wasm-type-impl.h`?

Reply:
Interesting idea! It sounds like it would require at least 3 headers, though:

1. One header for HeapType.
2. One header for TypeInfo, which depends on HeapType (even a forward declaration is insufficient).
3. One header for the rest of wasm-type.h

I'm not sure if that much splitting is worth it?

Reply:
oof, nope. Scratch that idea.